src/environmentd/src/http.rs

```diff
@@ -188,62 +188,93 @@ pub struct InternalHttpConfig {
     pub active_connection_count: Arc<Mutex<ConnectionCounter>>,
-    pub promote_leader: Option<oneshot::Sender<()>>,
-    pub ready_to_promote: Option<oneshot::Receiver<()>>,
+    pub promote_leader: oneshot::Sender<()>,
+    pub ready_to_promote: oneshot::Receiver<()>,
 }
 
 pub struct InternalHttpServer {
     router: Router,
 }
 
-#[derive(Debug)]
-pub struct LeaderState {
-    status: LeaderStatus,
-    pub promote_leader: Option<oneshot::Sender<()>>,
-    pub ready_to_promote: Option<oneshot::Receiver<()>>,
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct LeaderStatusResponse {
+    pub status: LeaderStatus,
 }
 
-#[derive(Debug, Serialize, Deserialize)]
-struct LeaderStatusResponse {
-    status: LeaderStatus,
+#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum LeaderStatus {
+    /// The current leader returns this. It shouldn't need to know that another instance is attempting to update it.
+    IsLeader,
+    /// The new pod should return this status until it is ready to become the leader, or it has determined that it cannot proceed.
+    Initializing,
+    /// Once we receive this status, we can tell it to become the leader and migrate the EIP.
+    ReadyToPromote,
 }
 
-#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
-enum LeaderStatus {
-    IsLeader, // 200: The current leader returns this. It shouldn't need to know that another instance is attempting to update it.
-    Initializing, // 200: The new pod should return this status until it is ready to become the leader, or it has determined that it cannot proceed.
-    ReadyToPromote, // 200: Once we receive this status, we can tell it to become the leader and migrate the EIP.
+#[derive(Debug)]
+pub enum LeaderState {
+    IsLeader,
+    Initializing {
+        promote_leader: oneshot::Sender<()>,
+        ready_to_promote: oneshot::Receiver<()>,
+    },
+    ReadyToPromote {
+        promote_leader: oneshot::Sender<()>,
+    },
+}
+
+fn state_to_status(state: &LeaderState) -> LeaderStatus {
+    match state {
+        LeaderState::IsLeader => LeaderStatus::IsLeader,
+        LeaderState::Initializing { .. } => LeaderStatus::Initializing,
+        LeaderState::ReadyToPromote { .. } => LeaderStatus::ReadyToPromote,
+    }
 }
 
 pub async fn handle_leader_status(
     State(state): State<Arc<Mutex<LeaderState>>>,
 ) -> impl IntoResponse {
     let mut leader_state = state.lock().expect("lock poisoned");
-    if let Some(mut ready_to_promote) = leader_state.ready_to_promote.take() {
-        assert_eq!(leader_state.status, LeaderStatus::Initializing);
-        match ready_to_promote.try_recv() {
-            Ok(()) => leader_state.status = LeaderStatus::ReadyToPromote,
-            Err(TryRecvError::Empty) => {
-                leader_state.ready_to_promote = Some(ready_to_promote);
+    let mut state = LeaderState::IsLeader;
+    std::mem::swap(&mut *leader_state, &mut state);
```

Comment:
Hrm this is still awkward. Is there some rust borrow reason it needs to either swap or do the Option take from before?

Reply:
Yeah. oneshot::{Sender, Receiver} aren't copy/clone and Sender::send takes `self` so we need to move it out of the mutex. It is awkward especially when we sometimes have to move it back into the mutex (like the Initializing case).